R/jointRPCA.R

```diff
@@ -0,0 +1,752 @@
+#'   that defines training vs test samples.
+#' @param n.components Integer specifying the number of principal components to compute.
+#' @param rclr.transform.tables Logical; whether to apply rCLR transformation to each
+#'   input table before ordination. Default is \code{TRUE}.
+#' @param min.sample.count Minimum total count required for a sample to be retained.
+#' @param min.feature.count Minimum total count required for a feature to be retained.
+#' @param min.feature.frequency Minimum percentage (0–100) of samples in which a
+#'   feature must be non-zero to be retained.
+#' @param max.iterations Maximum number of optimization iterations.
+#'
+#' @return A list with \code{ord.res}, \code{dist}, \code{cv.stats}, and
+#'   \code{rclr.tables}.
+#'
+#' @keywords internal
+#' @noRd
+
+.jointRPCA <- function(tables,
+                      n.test.samples = 10,
+                      sample.metadata = NULL,
+                      train.test.column = NULL,
+                      n.components = 3,
+                      rclr.transform.tables = TRUE,
+                      min.sample.count = 0,
+                      min.feature.count = 0,
+                      min.feature.frequency = 0,
+                      max.iterations = 5) {
+    
+    if (is.null(names(tables))) {
+        names(tables) <- paste0("view", seq_along(tables))
+    }
+    
+    if (n.components < 2) stop("n.components must be at least 2.")
+    if (max.iterations < 1) stop("max.iterations must be at least 1.")
+    
+    #filtering (always done, independent of rclr)
+    tables <- lapply(tables, function(tbl) {
+        .rpca_table_processing(
+            tbl,
+            min.sample.count      = min.sample.count,
+            min.feature.count     = min.feature.count,
+            min.feature.frequency = min.feature.frequency
+        )
+    })
+    
+    #find shared samples across views
+    sample.sets <- lapply(tables, colnames)
+    shared.all.samples <- Reduce(intersect, sample.sets)
+    if (length(shared.all.samples) == 0) {
+        stop("No samples overlap between all tables. If using pre-transformed tables, set rclr.transform.tables = FALSE.")
+    }
+    unshared.samples <- setdiff(unique(unlist(sample.sets)), shared.all.samples)
+    if (length(unshared.samples) > 0) {
+        warning(sprintf("Removing %d sample(s) that do not overlap in tables.", length(unshared.samples)))
+    }
+    
+    #restrict each table to the shared sample set
+    tables <- lapply(tables, function(tbl) {
+        tbl[, shared.all.samples, drop = FALSE]
+    })
+    shared.all.samples <- Reduce(intersect, lapply(tables, colnames))
+    
+    #transform tables: rCLR or masking
+    rclr.tables <- lapply(tables, function(tbl) {
+        mat <- as.matrix(tbl)
+        if (rclr.transform.tables) {
+            vegan::decostand(mat, method = "rclr", MARGIN = 2)
+        } else {
+            .mask_value_only(mat)$data
+        }
+    })
+    names(rclr.tables) <- names(tables)
```

Comment:
Data transformation should be done outside the function. I think user should be in full control and the function should do only what it says that it is doing.